### PR TITLE
TOOLS-1941 TOOLS-2200 Add namespace level xdr stats and add xdr filter command

### PR DESCRIPTION
**Pull request by jdogmcsteezy**
Added 5 new commands: `show config xdr dc`, `show config xdr namespace`, `show config xdr filter`, `show stat xdr dc`, `show stat xdr namespace`.

The `show config xdr` and `show stat xdt` commands run their `dc` and `namespace` children still. The deprecated `show config dc` command is now deprecated for all server version. Before, the deprecation warning was only displayed for pre xdr5 versions. 

The data model returned from a node/cluster when asking for the xdr configs has been changed.  You will see some translation code in the log_handler to handle this. The model has been re-architected in such a way that checks for pre vs post xdr5 version is no longer necessary.

### CODE REVIEW

File: lib/collectinfo_analyzer/show_controller.py [hunk 1]
```diff
@@ -32,21 +36,29 @@ def _do_default(self, line):
...
     )
     def _do_default(self, line):
         self.do_security(line[:])
         self.do_service(line[:])
         self.do_network(line[:])
         self.do_namespace(line[:])
 
-    @CommandHelp("Displays service configuration")
+    @CommandHelp(
+        "Displays service configuration",
```
Comment (dwelch-spike) [inline lib/collectinfo_analyzer/show_controller.py:56]: Should this help text be something about security?

File: lib/collectinfo_analyzer/show_controller.py [hunk 2]
```diff
@@ -281,49 +381,112 @@ def do_dc(self, line):
...
-                        "%s DC Configuration" % (dc),
-                        configs,
-                        self.log_handler.get_cinfo_log_at(timestamp=timestamp),
-                        title_every_nth=title_every_nth,
-                        flip_output=flip_output,
-                        timestamp=timestamp,
-                        **self.mods
-                    )
+    @CommandHelp(
+        "Displays xdr namespace configuration.",
```
Comment (dwelch-spike) [inline lib/collectinfo_analyzer/show_controller.py:450]: Should the help text be about XDR filters?

File: lib/collectinfo_analyzer/show_controller.py [hunk 3]
```diff
@@ -916,17 +1051,204 @@ def do_sindex(self, line):
...
+        "    -r           - Repeat output table title and row header after every <terminal width> columns.",
+        "                   [default: False, no repetition]",
+        "    --flip       - Flip output table to show Nodes on Y axis and stats on X axis.",
+    )
+    def _do_default(self, line):
+        self._do_xdr(line[:])
+        self.do_dc(line[:])
+        self.do_namespace(line[:])
+
+    def _do_xdr(self, line):
```
Comment (dwelch-spike) [inline lib/collectinfo_analyzer/show_controller.py:1095]: No help text?

File: lib/collectinfo_analyzer/show_controller.py [hunk 4]
```diff
@@ -916,17 +1051,204 @@ def do_sindex(self, line):
...
+        )
+
+        xdr_stats = self.getter.get_xdr()
+
+        for timestamp in xdr_stats.keys():
+            cinfo_log = self.log_handler.get_cinfo_log_at(timestamp=timestamp)
+
+            # There are no XDR level stats for XDR 5. This will not print anything
+            # if it is empty
+            self.view.show_stats(
```
Comment (dwelch-spike) [inline lib/collectinfo_analyzer/show_controller.py:1129]: Just checking to make sure there isn't a specific view function for this. show_stats() is ok?

File: lib/collectinfo_analyzer/show_controller.py [hunk 5]
```diff
@@ -916,17 +1051,204 @@ def do_sindex(self, line):
...
+            mods=self.mods,
+        )
+
+        by_dc = util.check_arg_and_delete_from_mods(
+            line=line,
+            arg="--by-dc",
+            default=False,
+            modifiers=self.modifiers,
+            mods=self.mods,
+        )
```
Comment (dwelch-spike) [inline lib/collectinfo_analyzer/show_controller.py:1231]: Does the `--by-dc` modifier need help text?

**Review by dwelch-spike — CHANGES_REQUESTED** [review]
Mostly wondering about the get_controller2.py file. Is it needed? 

Since we did the rest of the code in a review together I only went through the files under collectinfo_analyzer

File: lib/collectinfo_analyzer/collectinfo_handler/collectinfo_log.py [hunk 6]
```diff
@@ -235,6 +236,40 @@ def get_data(self, type="", stanza=""):
...
+                        design. That is why this is here.
+                        """
+                        if "xdr_configs" in xdr_stats and stanza == "xdr":
+                            data[node] = copy.deepcopy(xdr_stats["xdr_configs"])
+                        elif "dc_configs" in xdr_stats and stanza == "dc":
+                            data[node] = copy.deepcopy(xdr_stats["dc_configs"])
+                        elif "ns_configs" in xdr_stats and stanza == "xdr_ns":
+                            data[node] = copy.deepcopy(xdr_stats["ns_configs"])
+                        else:
+                            data[node] = copy.deepcopy(d[stanza])
```
[the remark this answers is not in the document]
Comment (jdogmcsteezy) [reply lib/collectinfo_analyzer/collectinfo_handler/collectinfo_log.py:255]: A `key_error` here would indicate a developer mistake

File: lib/collectinfo_analyzer/get_controller.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+                dcs = dc_vals.keys()
+                filtered_dcs = set(util.filter_list(dcs, dcs_filter))
+
+                for dc, ns_filters in list(dc_vals.items()):
+                    if dc not in filtered_dcs:
+                        del dc_vals[dc]
+                        continue
+
+                    namespaces = ns_filters.keys()
```
[the remark this answers is not in the document]
Comment (jdogmcsteezy) [reply lib/collectinfo_analyzer/get_controller.py:85]: good call.  It was copied from some where that the variable name made sense

File: lib/collectinfo_analyzer/show_controller.py (hunk 3 above)
Comment (jdogmcsteezy) [reply lib/collectinfo_analyzer/show_controller.py:1095]: no, not a command that can be called

File: lib/collectinfo_analyzer/show_controller.py (hunk 5 above)
Comment (jdogmcsteezy) [reply lib/collectinfo_analyzer/show_controller.py:1231]: yes, good catch

File: lib/collectinfo_analyzer/show_controller.py (hunk 2 above)
Comment (jdogmcsteezy) [reply lib/collectinfo_analyzer/show_controller.py]: yes

**Pull request merged**